wallet-new/src/Cardano/Wallet/API/Types.hs

```diff
@@ -11,13 +11,146 @@ import           Universum
+        :: forall n someApiType. (KnownSymbol n, someApiType n ~ apiType)
+        => Proxy (someApiType n a) -> String
+    apiArgName _ = symbolVal (Proxy @n)
+
+type ApiHasArgInvariant apiType a res =
+    Server (apiType a :> res) ~ (ApiArg apiType a -> Server res)
+
+type ApiHasArg apiType a res =
+    ( ApiHasArgClass apiType a
+    , ApiHasArgInvariant apiType a res
+    )
+
+instance KnownSymbol s => ApiHasArgClass (Capture s) a
+instance KnownSymbol s => ApiHasArgClass (QueryParam s) a where
+    type ApiArg (QueryParam s) a = Maybe a
+instance KnownSymbol s => ApiHasArgClass (QueryParams s) a where
+    type ApiArg (QueryParams s) a = [a]
+instance KnownSymbol s => ApiHasArgClass (Header s) a where
+    type ApiArg (Header s) a = Maybe a
+instance ApiHasArgClass (ReqBody ct) a where
+    apiArgName _ = "request body"
+
+-- | `Unmaybe` and `UnmaybeArg` type families are necessary to
+-- get `a` from `Maybe a` in a way which doesn't make GHC mad.
+-- Used in definition of `ApiHasArgClass` and `HasServer` instances for
+-- `WithDefaultApiArg`
+type family Unmaybe a where
+    Unmaybe (Maybe a) = a
+
+type family UnmaybeArg a where
+    UnmaybeArg (Maybe a -> b) = a -> b
+
+--
+-- Specifying API method parameters with default values
+--
+
+-- | `WithDefaultApiArg` type is used to specify API parameters,
+-- which are optional but have default values. It can be applied to every
+-- API specifier which yields `Maybe a` as argument type.
+data WithDefaultApiArg (argType :: * -> *) a
+
+instance (ApiHasArgClass apiType a, ApiArg apiType a ~ Maybe b) =>
+         ApiHasArgClass (WithDefaultApiArg apiType) a where
+    type ApiArg (WithDefaultApiArg apiType) a = Unmaybe (ApiArg apiType a)
+    apiArgName _ = apiArgName (Proxy @(apiType a))
+
+instance ( HasServer (apiType a :> res) ctx
+         , Server (apiType a :> res) ~ (Maybe c -> d)
+         , Default c
+         ) =>
+         HasServer (WithDefaultApiArg apiType a :> res) ctx where
+    type ServerT (WithDefaultApiArg apiType a :> res) m =
+        UnmaybeArg (ServerT (apiType a :> res) m)
+    route =
+        inRouteServer @(apiType a :> res) route $
+        \f a -> f $ fromMaybe def a
+
+instance HasSwagger (apiType a :> res) =>
+    HasSwagger (WithDefaultApiArg apiType a :> res) where
+    toSwagger _ = toSwagger (Proxy @(apiType a :> res))
+
+-- | Type aliases for query params and headers which have default values.
+type DQueryParam s a = WithDefaultApiArg (QueryParam s) a
+type DHeader s a = WithDefaultApiArg (Header s) a
+
+--
+-- Providing one argument via different methods
+--
+type family MergeArgs a where
+    MergeArgs (a -> a -> b) = a -> b
+
+-- | `AlternativeApiArg` is used when we want to provide a way
+-- to pass one argument via two ways, e. g. via request header or query parameter.
+data AlternativeApiArg (argA :: * -> *) (argB :: * -> *) a
+
+instance ( ApiHasArgClass argA a
+         , ApiHasArgClass argB a
+         , ApiArg argA a ~ Maybe b
```

Comment:
Extra constraint?

Reply:
Because we need to choose between two parameters, and we cannot meaningfully do it if both of them are always present (we can always choose the first one, but this doesn't make sense).
But thanks for the comment, actually - I realized that first constraint is unnecessary, and constraint in `HasServer` can be generalized to requiring any  `Alternative`.